src/ui/Sort/SortCriteria.ts

```diff
@@ -1,37 +1,91 @@
+import { chain, contains, first, values } from 'underscore';
 import { Assert } from '../../misc/Assert';
-import { QueryBuilder } from '../Base/QueryBuilder';
-import { Utils } from '../../utils/Utils';
-import { IQueryResult } from '../../rest/QueryResult';
-import * as _ from 'underscore';
+import { QueryBuilder } from '../../ui/Base/QueryBuilder';
 
-export class SortCriteria {
-  private static validSorts = ['relevancy', 'date', 'qre'];
-  private static sortsNeedingDirection = ['date'];
-  private static validDirections = ['ascending', 'descending'];
+export enum VALID_SORT {
+  RELEVANCY = 'relevancy',
+  DATE = 'date',
+  QRE = 'qre'
+}
+
+export enum VALID_DIRECTION {
+  ASCENDING = 'ascending',
+  DESCENDING = 'descending'
+}
+
+export class SingleSortCriteria {
+  private static sortsNeedingDirection = [VALID_SORT.DATE];
 
   /**
    * Create a new SortCriteria
    * @param sort The sort criteria (e.g.: relevancy, date)
    * @param direction The direction by which to sort (e.g.: ascending, descending)
    */
-  constructor(public sort: string, public direction: string = '') {
-    Assert.isNonEmptyString(sort);
-    Assert.check(_.contains(SortCriteria.validSorts, sort) || SortCriteria.sortIsField(sort));
-    if (SortCriteria.sortNeedsDirection(sort)) {
-      Assert.check(_.contains(SortCriteria.validDirections, direction));
+  constructor(public sort: VALID_SORT, public direction: VALID_DIRECTION | '' = '') {
+    if (!SingleSortCriteria.sortIsField(sort)) {
+      Assert.check(
+        this.isValidSort(sort),
+        `${sort} is not a valid sort criteria. Valid values are ${values(VALID_SORT)} or a valid index sortable index field.`
+      );
+    }
+    if (SingleSortCriteria.sortNeedsDirection(sort)) {
+      Assert.check(
+        this.isValidDirection(direction),
+        `${direction} is not a valid sort criteria direction. Valid values are ${values(VALID_DIRECTION)}`
+      );
     } else {
       Assert.check(direction == '');
     }
   }
 
+  private isValidDirection(direction: string): direction is VALID_DIRECTION {
+    return chain(VALID_DIRECTION)
+      .values()
+      .contains(direction as any)
```

Comment:
Was it necessary to type `direction` as any here?

Similarly, `sort` on line 51.

Reply:
Yes, because of the type definition for underscore JS + TypeScript String enum clashes a bit.

`chain<VALID_DIRECTION>.values -> VALID_DIRECTION[].contains(foo)`, foo must be a VALID_DIRECTION, but here I want to check if an arbitrary string is a VALID_DIRECTION